bot/core/paginator.py

```diff
@@ -0,0 +1,421 @@
+        paginator : t.Union[ImagePaginator, ListPaginator, EmbedPaginator, FieldPaginator]
+            The paginator itself.
+            This can be an image, list, embed or a field paginator.
+            Note that it MUST be one of those.
+            If you want to process a normal paginator from discord.py,
+            use this:
+
+            paginator = ListPaginator(entries=['Hello World!'], individual_paging=False)
+            await process_paginator(paginator, ...)
+
+        context: t.Union[commands.Context, discord.Message]
+            The context, used for sending a message to the channel, getting the author etc.
+
+        bot: commands.Bot
+            Used to run 'bot.wait_for' and other things
+        show_page_length: bool
+            Whether to show the page ('Page 1/4') in the footer
+        timeout: float
+            For how long to respond to a new reaction.
+            If this time limit is reached, all of the reactions clear
+            and the function exits, breaking the loop.
+        **kwargs: t.Optional[t.Union[discord.Emoji, str]]
+            Used to override reaction emojis.
+            For example, process_paginator(..., first_page_emoji=😂'<:emoji:192391292319>')
+            This will replace the emoji that indicates the first page (by default
+            set to ⏮) with the emoji that we gave to that kwarg.
+    """
+    if not check_permissions(context.channel, bot):
+        raise PaginatorError("Bot lacks a permission to paginate")
+
+    supported_types = (ImagePaginator, ListPaginator, EmbedPaginator, FieldPaginator)
+
+    if isinstance(context, Context):
+        send = context.send
+    elif isinstance(context, discord.Message):
+        send = context.channel.send
+    else:
+        raise PaginatorError(
+            "The context argument given to process_paginator "
+            "must be a commands.Context or a discord.Message object."
+        )
+
+    if not type(paginator) in supported_types:
+        raise PaginatorError(dedent(
+            "The paginator must be an instance of one of these:\n"
+            f"{', '.join([type_.__name__ for type_ in supported_types])}\n"
+            "If you want to process commands.Paginator, try out "
+            "ListPaginator like so:\n"
+            "Paginator = ListPaginator(['Hello World'])"
+        ))
+
+    page = 0
+
+    embed = paginator.embeds[page]
+
+    # Reaction emojis. Can be overridden by kwargs.
+    reaction_emojis = {
+        "first_page_emoji": "⏮️",  # First page
+        "prev_page_emoji": "⬅️",  # Previous Page
+        "next_page_emoji": "➡️",  # Next Page
+        "last_page_emoji": "⏭️",  # Last Page
+        "exit_emoji": "<:trashcan:722035312238526505>",  # Stop paginator and delete message (trashcan)
+    }
+
+    for key in reaction_emojis.keys():
+        if key in [_ for _ in kwargs]:
+            reaction_emojis[key] = kwargs[key]
+
+    def update_footer(embed_: discord.Embed, page_: int) -> None:
```

Comment:
What's wrong with using `embed` and `page` why `embed_` and `page_`?

**Same applies for all functions with seemingly irrelevant underscores next to their arguments**
```suggestion
    def update_footer(embed: discord.Embed, page: int) -> None:
```